lib/Dialect/StableHLO/Transforms/ReoutlineComposite.cpp

```diff
@@ -0,0 +1,373 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "ttmlir/Dialect/StableHLO/Transforms/Passes.h"
+#include "ttmlir/Dialect/StableHLO/Utils/ShardingUtils.h"
+#include "ttmlir/Dialect/StableHLO/Utils/ShardyUtils.h"
+#include "llvm/Support/Error.h"
+
+namespace mlir::tt::stablehlo {
+#define GEN_PASS_DEF_REOUTLINECOMPOSITEPASS
+#include "ttmlir/Dialect/StableHLO/Transforms/Passes.h.inc"
+
+using GroupKey = mlir::StringAttr;
+
+// Collect ops per group string inside a func.
+static void collectGroups(
+    mlir::func::FuncOp func,
+    llvm::DenseMap<GroupKey, llvm::SmallVector<mlir::Operation *>> &groups) {
+  func.walk([&](mlir::Operation *op) {
+    if (auto attr =
+            op->getAttrOfType<mlir::StringAttr>(sharding_utils::kGroupAttr)) {
+      groups[attr].push_back(op);
+    }
+  });
+}
+
+static bool isOpBefore(mlir::Operation *opA, mlir::Operation *opB) {
+  if (opA->getBlock() != opB->getBlock()) {
+    return false;
+  }
+  return opA->isBeforeInBlock(opB);
+}
+
+// Sort ops by dominance order within a single block (we require one block
+// later).
+static void sortByBlockOrder(llvm::SmallVector<mlir::Operation *> &ops) {
+  llvm::sort(ops, isOpBefore);
+}
+
+// Compute captured inputs (defined outside, used by the group) and escaping
+// outputs (defined in the group, used outside). Also verify the group is
+// inside a single block and forms a contiguous range (movable as a chunk).
+static bool analyzeBoundary(const llvm::SmallVector<mlir::Operation *> &ops,
+                            llvm::SmallVector<mlir::Value> &captures,
+                            llvm::SmallVector<mlir::Value> &escapes,
+                            mlir::Operation *&firstOp,
+                            mlir::Operation *&lastOp) {
+  // Sort and find contiguous range [first, last].
+  llvm::SmallVector<mlir::Operation *> sorted = ops;
+  sortByBlockOrder(sorted);
+  firstOp = sorted.front();
+  lastOp = sorted.back();
+
+  // Build a set for quick membership tests.
+  llvm::SmallDenseSet<mlir::Operation *> inGroup;
+  inGroup.reserve(sorted.size());
+  for (auto *op : sorted) {
+    inGroup.insert(op);
+  }
+
+  for (mlir::Operation &it :
+       llvm::make_range(mlir::Block::iterator(firstOp),
+                        std::next(mlir::Block::iterator(lastOp)))) {
+    // All ops between first and last must be in the group.
+    // If new ops were inserted in between, we cannot reoutline.
+    if (!inGroup.contains(&it)) {
+      return false;
+    }
+  }
+
+  // Captures: operands whose defining op is not in the group (or is a block
+  // arg).
+  llvm::SmallDenseSet<mlir::Value> captureSet;
+  for (auto *op : sorted) {
+    for (mlir::Value v : op->getOperands()) {
+      if (auto defOp = v.getDefiningOp()) {
```

Comment:
:warning: **llvm\-qualified\-auto** :warning:
`` auto defOp `` can be declared as `` auto *defOp ``
```suggestion
      if (auto *defOp = v.getDefiningOp()) {
```